### PR TITLE
Fixes crash in deformable convolutions (2598)

### LINKED ISSUE #2598: ops.deform_conv2d causes CUDA illegal memory access
## 🐛 Bug
I try to test the speed of deformable conv2d. But always encountered memory error.

## To Reproduce

```
$ ipython
Python 3.8.5 (default, Jul 27 2020, 08:42:51) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.17.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch
   ...: import torchvision as tv
   ...: weight = torch.randn(9,9,3,3).cuda()
   ...: weight.requires_grad = True
   ...: img = torch.randn(8, 9, 1000, 110).cuda()
   ...: def test():
   ...:     offset = torch.randn(8,18,1000,110).cuda()
   ...:     out = tv.ops.deform_conv2d(img, offset, weight, padding=1)
   ...:     out.mean().backward()
   ...: 

In [2]: import os

In [3]: os.environ["CUDA_LAUNCH_BLOCKING"] = "1"

In [4]: timeit test()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-a1086d7a4706> in <module>
----> 1 get_ipython().run_line_magic('timeit', 'test()')

/usr/lib/python3.8/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2324                 kwargs['local_ns'] = self.get_local_scope(stack_depth)
   2325             with self.builtin_trap:
-> 2326                 result = fn(*args, **kwargs)
   2327             return result
   2328 

<decorator-gen-60> in timeit(self, line, cell, local_ns)

/usr/lib/python3.8/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188 
    189         if callable(arg):

/usr/lib/python3.8/site-packages/IPython/core/magics/execution.py in timeit(self, line, cell, local_ns)
   1171                     break
   1172 
-> 1173         all_runs = timer.repeat(repeat, number)
   1174         best = min(all_runs) / number
   1175         worst = max(all_runs) / number

/usr/lib/python3.8/timeit.py in repeat(self, repeat, number)
    203         r = []
    204         for i in range(repeat):
--> 205             t = self.timeit(number)
    206             r.append(t)
    207         return r

/usr/lib/python3.8/site-packages/IPython/core/magics/execution.py in timeit(self, number)
    167         gc.disable()
    168         try:
--> 169             timing = self.inner(it, self.timer)
    170         finally:
    171             if gcold:

<magic-timeit> in inner(_it, _timer)

<ipython-input-1-a97200bb984a> in test()
      5 img = torch.randn(8, 9, 1000, 110).cuda()
      6 def test():
----> 7     offset = torch.randn(8,18,1000,110).cuda()
      8     out = tv.ops.deform_conv2d(img, offset, weight, padding=1)
      9     out.mean().backward()

RuntimeError: CUDA error: an illegal memory access was encountered

In [5]: 
```

## Environment
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 11.0

OS: Arch Linux (x86_64)
GCC version: (GCC) 10.1.0
Clang version: 10.0.1 
CMake version: version 3.18.1

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.0.2
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti
Nvidia driver version: 450.57
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.2
/usr/lib/libcudnn_adv_infer.so.8.0.2
/usr/lib/libcudnn_adv_train.so.8.0.2
/usr/lib/libcudnn_cnn_infer.so.8.0.2
/usr/lib/libcudnn_cnn_train.so.8.0.2
/usr/lib/libcudnn_ops_infer.so.8.0.2
/usr/lib/libcudnn_ops_train.so.8.0.2

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torch-cluster==1.4.5
[pip3] torch-geometric==1.3.2
[pip3] torch-scatter==1.4.0
[pip3] torch-sparse==0.4.3
[pip3] torchvision==0.7.0a0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #2598

Description:

~~[x] Adjusted `num_kernels` and `batch_size` according to `kMaxGridNum * CUDA_NUM_THREADS`~~
- [x] Redefined `kMaxGridNum` as max grid according to current CUDA device
- [x] Added test to check the code from issue and compared grads CPU/CUDA

~~This PR adjusts `num_kernels` and `batch_size` to avoid the situation such that `num_kernels` > `kMaxGridNum * CUDA_NUM_THREADS`. If it is not possible to adjust `batch_size` (e.g. batch_size=0), an exception is raised.~~

### CODE REVIEW

File: torchvision/csrc/cuda/DeformConv_cuda.cu [hunk 1]
```diff
@@ -80,12 +80,11 @@
 
 using namespace at;
 
-const int CUDA_NUM_THREADS = 1024;
-const int kMaxGridNum = 65535;
-
+const unsigned int CUDA_NUM_THREADS = 1024;
+const unsigned int kMaxGridNum = at::cuda::getCurrentDeviceProperties()->maxGridSize[0];
```
Comment (fmassa) [inline torchvision/csrc/cuda/DeformConv_cuda.cu:84]: For my understanding, can we make this a `const` in the outer namespace? Doesn't it depend on the machine it's being run on, and thus could change from the machine it was compiled to the machine it's being run on?

**Review by fmassa — COMMENTED** [review]
Looks great, thanks a lot!

I have one comment (which might be silly), let me know what you think

File: torchvision/csrc/cuda/DeformConv_cuda.cu (hunk 1 above)
Comment (vfdev-5) [reply torchvision/csrc/cuda/DeformConv_cuda.cu:84]: Yes, I agree that this variable should be assigned at runtime. It is very unlikely that taking into account how implemented [`at::cuda::getCurrentDeviceProperties()`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAContext.cpp#L41), that `kMaxGridNum` is evaluated at compilation time.

Comment (vfdev-5) [reply torchvision/csrc/cuda/DeformConv_cuda.cu]: Yes, you are right. I'll change it to dynamic evaluation

Comment (vfdev-5) [reply torchvision/csrc/cuda/DeformConv_cuda.cu]: Yes, you are right. I'll change it to dynamic evaluation

**Review by fmassa — APPROVED** [review]
Thanks a lot @vfdev-5 !

**Pull request merged**